### DISCUSSION TITLE
node v20 --loader error

### DISCUSSION THREAD

**Discussion by iambumblehead** [avajs/ava#3199] [Q&A]
Checking in here to notify about an issue using ava with node v20 and --loader. There is a problem demonstrated at the repo here: https://github.com/iambumblehead/nodejs-v20-loader-ava-noop; using `npm test` from that repo with node v20 causes the test to incorrectly fail.

Essentially ava fails when used with --loader

_anyloader.js_
```javascript
export const load = async (url, context, nextLoad) => {
  if (url.endsWith('anyfile.js')) {
    return {
      format: 'module',
      shortCircuit: true,
      responseURL: url,
      source: 'export default "fromloader"'
    }
  }
  return nextLoad(url, context)
}
```

_anyfile.spec.js_
```javascript
import test from 'ava'

test('should user loader', async t => {
  const anyfile = await import('./anyfile.js')

  t.is(anyfile.default, 'fromloader')
})
```

```console
ava --node-arguments="--loader=./anyloader.js"


(node:59341) ExperimentalWarning: Custom ESM Loaders is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
  ✖ should user loader Promise returned by test never resolved
  ─

  should user loader

  Error: Promise returned by test never resolved

  ─

  1 test failed
```

It might be caused by an upstream issue reported earlier this evening https://github.com/nodejs/node/issues/47614, however opening an issue here may be useful because the error manifests itself in a special way with ava, indicating a possibly separate issue.

**Comment by novemberborn** [comment]
I'd argue that custom loaders should not break AVA if Node.js and the loader are behaving correctly.

**Reply by iambumblehead** [reply]
Using the latest node 20.2.0, only ava does not work with the loaders, other test-runners are working: uvu, node:test, mocha, jest and various combinations thereof using tsm and ts-node/esm.

**Reply by novemberborn** [reply]
You'll have to do some spelunking to see what AVA is doing that is breaking it then. I don't think we do anything special that should be impacted by normally behaving loaders 🤔
